### PR TITLE
MRG: add `kmers_and_hashes` method to get canonical kmers + hashes

### PR DESCRIPTION
**Pull request by ctb**
Tackles https://github.com/oxli-bio/oxli/issues/21.

This PR:
* makes `hash_kmer` accessible to Python;
* provides a method `kmers_and_hashes` that returns a list of `(canonical_kmer, hashval)` tuples without modifying the HashMap;
* adds appropriate tests.

### CODE REVIEW

**Comment by ctb** [comment]
@Adamtaranto I am not attached to this code or implementation. Please feel free to munge at will.

File: src/lib.rs
Comment (Adamtaranto) [inline src/lib.rs:400]: Should we set this with `allow_bad_kmers` ?

File: src/lib.rs
Comment (Adamtaranto) [inline src/lib.rs:411]: Set `substr` to be the canonical form of the kmer.

File: src/lib.rs
Comment (Adamtaranto) [inline src/lib.rs:413]: Check what happens here if `allow_bad_kmers` is False and a bad kmer is encountered?

Is hashval null? error?

File: src/lib.rs
Comment (Adamtaranto) [reply src/lib.rs:411]: i.e. make uppercase, get revcomp, get min of kmer and revkmer.

File: src/lib.rs
Comment (Adamtaranto) [inline src/lib.rs:434]: Only return (kmer, hash) if valid hash was generated.

File: src/lib.rs
Comment (Adamtaranto) [reply src/lib.rs:413]: Can hashes get out of sync with substr slices if a bad kmer is skipped?

Comment (ctb) [reply src/lib.rs:413]: this is where the rust error handling is nice - unwrap() will just panic! and then there are ways of handling the rest when we get to it.

I'll write various tests to make sure we cover the cases.

**Comment by Adamtaranto** [comment]
@ctb what about making the [`kmer_index`](https://github.com/sourmash-bio/sourmash/blob/d63c464e825529fa54bb7e8b81faa53b858b09de/src/core/src/signature.rs#L216) attribute of SeqToHashes accessible via `SeqToHashes.get_kmer_index` ? 

something like: 

```rust
pub fn get_kmer_index(&self) -> usize {
        self.kmer_index
    }
```

That way we wouldn't have to worry about keeping seq slices and hashes in sync, we could just extract the seq slice from wherever the hasher is up to.

File: src/lib.rs
Comment (ctb) [reply src/lib.rs:565]: Returning ("", 0) for bad k-mers.

**Comment by ctb** [comment]
Ready for review! The code works and is well tested, I think, but could be improved in a few ways. Still it's at a good resting point so would like to suggest merging.

File: src/lib.rs
Comment (Adamtaranto) [inline src/lib.rs:515]: Is the full contig/chromosome in memory twice at this point?

File: src/lib.rs
Comment (Adamtaranto) [inline src/lib.rs:552]: Does sourmash have a canonicalize function we can recycle?

Else pop out as own function to make this cleaner?

File: src/lib.rs
Comment (Adamtaranto) [inline src/lib.rs:554]: It works, but I am uneasy about hoping that hasher output and seq slices stay in sync.

**Review by Adamtaranto — COMMENTED** [review]
Looks good. 

I'm in favour of outsourcing this to sourmash as a variant of `SeqToHashes` (`SeqToHashesAndKmers`?) So that we aren't doubling the seq copies in memory / looping over the seq twice.

**Comment by Adamtaranto** [comment]
I will try to add the following:

- Make kmers_and_hashes iterable
- Add hash2kmers hashmap
- opt to track kmers
- update consume to store kmers
- update count to store kmers

File: src/lib.rs
Comment (ctb) [reply src/lib.rs:552]: I didn't see anything in `SeqToHashes`, but will look again. Since that struct is old, and deals with more than just DNA k-mers, it may not have anything built in (and/or may not make it public).

File: src/lib.rs
Comment (ctb) [reply src/lib.rs:515]: no, it's a view (via the reference, `&`, see [as_bytes](https://doc.rust-lang.org/std/primitive.str.html#method.as_bytes)). 

If you calculate it, the memory consumption of the underlying string is minimal compared to catastrophic expansion of it into k-mers, even for human-size chromosomes.

**Comment by ctb** [comment]
If you are OK with the function signature, I'd suggest merging and then refactoring internally; generally my preference is to get the public API working and tested, and only then care about speed :). But your call!

File: src/lib.rs
Comment (ctb) [reply src/lib.rs:554]: (and to be clear, this is something I've wanted for sourmash for a while - right now we extract the hashes and kmers in Python, which is unbearably slow! So definitely want this in the sourmash library. Just need it to be pretty general, e.g. including proteins; and that codebase is also more complex to modify and requires approval on the PR, which takes time.)

**Review by Adamtaranto — APPROVED** [review]

**Comment by ctb** [comment]
ref https://github.com/sourmash-bio/sourmash/issues/3339

**Pull request merged**
